src/IECoreImage/ImageReader.cpp

```diff
@@ -51,23 +55,336 @@ using namespace IECoreImage;
 
 IE_CORE_DEFINERUNTIMETYPED( ImageReader );
 
-ImageReader::ImageReader( const std::string &description ) :
-		Reader( description, new ObjectParameter( "result", "The loaded object", new NullObject, ImagePrimitive::staticTypeId() ) )
+////////////////////////////////////////////////////////////////////////////////
+// ImageReader::Implementation
+////////////////////////////////////////////////////////////////////////////////
+
+class ImageReader::Implementation : public IECore::RefCounted
 {
-	m_dataWindowParameter = new Box2iParameter(
-		"dataWindow",
-		"The area for which data should be loaded. The default value (an empty box) "
-		"is used to specify that the full data window should be loaded. Other values may be specified "
-		"to load just a section of the image."
-	);
 
-	m_displayWindowParameter = new Box2iParameter(
-		"displayWindow",
-		"The displayWindow for the ImagePrimitive created during loading. The default value (an empty box) "
-		"is used to specify that the displayWindow should be inferred from the file itself. On rare occasions "
-		"it may be useful to specify an alternative using this parameter. Note that this parameter is completely "
-		"independent of the dataWindow parameter."
-	);
+	public :
+
+		Implementation( const ImageReader *reader ) : m_reader( reader ), m_inputFile( nullptr ), m_inputFileName( "" )
+		{
+		}
+
+		virtual ~Implementation()
+		{
+			ImageInput::destroy( m_inputFile );
+			m_reader = nullptr;
+		}
+
+		static bool canRead( const std::string &filename )
+		{
+			bool result = false;
+
+			if( ImageInput *input = ImageInput::open( filename ) )
```

Comment:
If I recall, `create()` was not doing the necessary thing (e.g. failing on invalid images) so I switched to `open()`. I can try switching back to verify my memory.